src/azul/portal_service.py

```diff
@@ -0,0 +1,167 @@
+from copy import deepcopy
+from typing import (
+    Callable,
+    Optional,
+)
+import json
+import logging
+
+import boto3
+
+from azul import config
+from azul.plugin import Plugin
+from azul.types import JSONs
+from azul.version_service import (
+    VersionService,
+    VersionConflict,
+    NoSuchObjectVersion,
+)
+
+log = logging.getLogger(__name__)
+
+
+class PortalService:
+
+    def __init__(self):
+        self.client = boto3.client('s3')
+        self.version_service = VersionService()
+
+    def crud(self, operation: Callable[[JSONs], Optional[JSONs]]) -> None:
+        """
+        Perform a concurrent read/write operation on the portal integrations DB.
+
+        :param operation: Callable that accepts the latest version of the portal
+        DB and optionally returns an updated version.
+        """
+        while True:
+            version = self.version_service.get(self._db_url())
+            if version is None:
+                log.info('Portal integration DB not found in S3; uploading hard-coded DB.')
+                self._create_db()
+                continue
```

Comment:
There is no "retry" for the create operation, which currently only happens when there's a request to operate on the DB but no DB exists (i.e. it is never explicitly requested via the API). If no DB exists at the time the operation is requested, but is created by another process before our process creates it, there is no need to re-create it (and trying to do so might actually erase data if a write operations sneaks in before our process manages to catch up).

The continuation is to ensure that we have the correct version before performing the requested operation. The `version` local is `None` before `_create_db` is called and must be updated to the current/initial version after the DB is created before `operation` can be performed. But we cannot correctly obtain the current version ID by returning it from `_create_db` because  while the DB will exist after this call it may have been created by another process. So we `continue` to re-read the current version at the top of the loop.